llvm/lib/Transforms/Utils/FunctionComparator.cpp

```diff
@@ -836,6 +837,54 @@ int FunctionComparator::cmpValues(const Value *L, const Value *R) const {
   return cmpNumbers(LeftSN.first->second, RightSN.first->second);
 }
 
+int FunctionComparator::cmpDistinctMetadata(const Function *L,
+                                            const Function *R) {
+  // Iterate over the basic blocks and instructions in Function L
+  for (const BasicBlock &BBL : *L) {
+    for (const Instruction &InstL : BBL) {
+      for (unsigned i = 0, e = InstL.getNumOperands(); i != e; ++i) {
+        if (MetadataAsValue *MDL =
+                dyn_cast<MetadataAsValue>(InstL.getOperand(i))) {
+          if (MDNode *MDNodeL = dyn_cast<MDNode>(MDL->getMetadata())) {
+            // Now iterate over the basic blocks and instructions in Function R
+            // to compare against each distinct metadata value in Function L
+            for (const BasicBlock &BBR : *R) {
+              for (const Instruction &InstR : BBR) {
+                for (unsigned j = 0, f = InstR.getNumOperands(); j != f; ++j) {
+                  if (MetadataAsValue *MDR =
+                          dyn_cast<MetadataAsValue>(InstR.getOperand(j))) {
+                    if (MDNode *MDNodeR =
+                            dyn_cast<MDNode>(MDR->getMetadata())) {
+                      // Compare distinct metadata
+                      if (MDNodeL->isDistinct() && MDNodeR->isDistinct()) {
```

Comment:
What if `MDNodeL == MDNodeR`? Should the functions be considered equal in that case?